drive/src/drive/flags.rs

```diff
@@ -30,44 +30,620 @@
 //! Flags
 //!
 
-use byteorder::{BigEndian, ReadBytesExt};
+use crate::drive::flags::StorageFlags::{
+    MultiEpoch, MultiEpochOwned, SingleEpoch, SingleEpochOwned,
+};
+use costs::storage_cost::removal::StorageRemovedBytes::SectionedStorageRemoval;
+use costs::storage_cost::removal::{StorageRemovalPerEpochByIdentifier, StorageRemovedBytes};
 use grovedb::ElementFlags;
+use integer_encoding::VarInt;
+use intmap::IntMap;
+use std::cmp::Ordering;
+use std::collections::BTreeMap;
 
-use crate::error::drive::DriveError;
+use crate::error::storage_flags::StorageFlagsError;
 use crate::error::Error;
 
+type EpochIndex = u16;
+
+type BaseEpoch = EpochIndex;
+
+type BytesAddedInEpoch = u32;
+
+type OwnerId = [u8; 32];
+
 /// Storage flags
-#[derive(Clone, Debug, Default, PartialEq)]
-pub struct StorageFlags {
-    /// Epoch
-    pub epoch: u16,
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum StorageFlags {
+    /// Single epoch
+    /// represented as byte 0
+    SingleEpoch(BaseEpoch),
+
+    /// Multi epoch
+    /// represented as byte 1
+    MultiEpoch(BaseEpoch, BTreeMap<EpochIndex, BytesAddedInEpoch>),
+
+    /// Single epoch owned
+    /// represented as byte 2
+    SingleEpochOwned(BaseEpoch, OwnerId),
+
+    /// Multi epoch owned
+    /// represented as byte 3
+    MultiEpochOwned(BaseEpoch, BTreeMap<EpochIndex, BytesAddedInEpoch>, OwnerId),
 }
 
 impl StorageFlags {
+    /// Create new single epoch storage flags
+    pub fn new_single_epoch(epoch: BaseEpoch, maybe_owner_id: Option<OwnerId>) -> Self {
+        match maybe_owner_id {
+            None => SingleEpoch(epoch),
+            Some(owner_id) => SingleEpochOwned(epoch, owner_id),
+        }
+    }
+
+    fn combine_owner_id<'a>(&'a self, rhs: &'a Self) -> Result<Option<&'a OwnerId>, Error> {
+        if let Some(our_owner_id) = self.owner_id() {
+            if let Some(other_owner_id) = rhs.owner_id() {
```

Comment:
`Option.contains` ?

Reply:
That is not stable. Also it doesn't work because rhs is another Option.